femtools/Particles.F90

```diff
@@ -57,33 +56,102 @@ module particles
+  ! One particle list for each subgroup
+  type(detector_linked_list), allocatable, dimension(:), save :: particle_lists
+  ! Timing info for group output
+  type(time_period_type), allocatable, dimension(:), save :: output_CS
 
-contains
+  !> Derived type to hold the number of scalar, vector and tensor attributes,
+  !! old attributes and old fields for a particle subgroup
+  type attr_counts_type
+    integer, dimension(3) :: attrs, old_attrs, old_fields
+  end type attr_counts_type
+
+  !> Derived type to hold scalar, vector and tensor attributes
+  type attr_vals_type
+    real, dimension(:), allocatable :: s
+    real, dimension(:,:), allocatable :: v
+    real, dimension(:,:,:), allocatable :: t
+  end type attr_vals_type
+
+  interface allocate
+    module procedure allocate_attr_vals
+  end interface allocate
 
+  interface deallocate
+    module procedure deallocate_attr_vals
+  end interface deallocate
+
+contains
+  !> Allocate an attr_vals_type structure, with the given number
+  !! of scalar, vector and tensor attributes, and the geometric dimension
+  !! of the problem.
+  subroutine allocate_attr_vals(vals, dim, counts)
+    !> Structure to allocate
+    type(attr_vals_type), pointer :: vals
+    !> Geometric dimension
+    integer, intent(in) :: dim
+    !> Counts of each rank of attribute
+    integer, dimension(3), intent(in) :: counts
+
+    allocate(vals)
+    allocate(vals%s(counts(1)))
+    allocate(vals%v(dim, counts(2)))
+    allocate(vals%t(dim, dim, counts(3)))
+  end subroutine allocate_attr_vals
+
+  !> Deallocate an attr_vals_type structure
+  subroutine deallocate_attr_vals(vals)
+    type(attr_vals_type), pointer :: vals
+
+    deallocate(vals%s)
+    deallocate(vals%v)
+    deallocate(vals%t)
+    deallocate(vals)
+  end subroutine deallocate_attr_vals
+
+  !> Initialise particles and set up particle file headers (per particle array)
   subroutine initialise_particles(filename, state, global, from_flredecomp, number_of_partitions)
-    !! Initialise particles and set up particle file headers (per particle array)
-    character(len = *), intent(in) :: filename
+    !> Experiment filename to prefix particle output files
+    character(len=*), intent(in) :: filename
+    !> Model state structure
     type(state_type), dimension(:), intent(in) :: state
-    logical, intent(in), optional :: global !! use global/parallel picker queries to determine particle elements
-    logical, intent(in), optional :: from_flredecomp !! change behaviour if we're being called by flredecomp
-    integer, intent(in), optional :: number_of_partitions !! number of processes to use for reading
+    !> Use global/parallel picker queries to determine particle elements?
+    logical, intent(in), optional :: global
+    !> Whether we're being called by flredecomp, this changes behaviour around
+    !! parallel access to particle files
+    logical, intent(in), optional :: from_flredecomp
```

Comment:
In general I prefer if logicals just say what they do, rather than "changes behaviour if called within a certain context" - if necessary with multiple logicals representing the different bits that are switched on or off. Fluidity used to be littered with this sort of context-specific logic that tends to propagate through different layers as well. Instead, every bit of code should make sense on its own.